File: src/hexa_kinematics/hexa_kinematics/leg_specs.py
"""LegSpec loader.

Expands ``hexa_description``'s ``geometry.yaml`` (which defines only the
two reference mounts ``l_front`` and ``l_middle``, with ``yaw_deg`` in
degrees) into the full six-leg dict by the URDF symmetry rules:

- rear mirrors front about the body y-axis: ``x → -x``, ``yaw → pi - yaw``.
- right mirrors left  about the body x-axis: ``y → -y``, ``yaw → -yaw``.

These match the ``mount_leg`` macro in ``hexapod.urdf.xacro``: the YAML
stays the single source of truth, and the URDF and this loader expand
it the same way.

The IK library's joint-zero convention (horizontal femur, tibia colinear
with femur) coincides with the URDF's joint-zero (each leg link extends
along its parent's +x at joint zero), so IK angles can be sent straight
to ros2_control without per-joint offsets. The physical servo zero is
the per-joint ``center`` returned by ``joint_config.load_joint_limits``
(loaded from the ``joints:`` block of ``geometry.yaml``); when the
real-hardware bridge lands it should subtract that center from the IK
angle before commanding each servo.
"""

import math
from pathlib import Path

import yaml

from .leg_geometry import LegSpec


LEG_NAMES: tuple[str, ...] = (
    "l_front",
    "l_middle",
    "l_rear",
    "r_front",
    "r_middle",
    "r_rear",
)


def load_leg_specs(geometry_yaml_path: str | Path) -> dict[str, LegSpec]:
    """Parse ``geometry.yaml`` and return one ``LegSpec`` per leg, by name.

    Segment lengths come from ``leg.*``; mount positions come from
    ``mounts.l_front`` / ``mounts.l_middle`` and are mirrored for the
    rear and right-side legs.
    """
    with open(geometry_yaml_path) as f:
        cfg = yaml.safe_load(f)
    leg_cfg = cfg["leg"]
    mounts = cfg["mounts"]

    coxa_len = float(leg_cfg["coxa_length"])
    femur_len = float(leg_cfg["femur_length"])
    tibia_len = float(leg_cfg["tibia_length"])

    front = mounts["l_front"]
    middle = mounts["l_middle"]

    out: dict[str, LegSpec] = {}
    for side in ("l", "r"):
        for name in ("front", "middle", "rear"):
            ref = middle if name == "middle" else front
            ref_yaw = math.radians(ref["yaw_deg"])
            x_fr = -ref["x"] if name == "rear" else ref["x"]
            yaw_fr = math.pi - ref_yaw if name == "rear" else ref_yaw
            mx = x_fr
            my = -ref["y"] if side == "r" else ref["y"]
            myaw = -yaw_fr if side == "r" else yaw_fr
            out[f"{side}_{name}"] = LegSpec(
                mount_xyz=(mx, my, 0.0),
                mount_yaw=myaw,
                coxa_len=coxa_len,
                femur_len=femur_len,
                tibia_len=tibia_len,
            )
    return out

File: src/hexa_kinematics_cpp/src/leg_specs.cpp
#include "hexa_kinematics_cpp/leg_specs.hpp"

#include <cmath>

#include "hexa_kinematics_cpp/yaml_util.hpp"

namespace hexa_kinematics {

std::map<std::string, LegSpec> load_leg_specs(
    const std::string& geometry_yaml_path) {
  const YAML::Node cfg = load_file(geometry_yaml_path);
  const YAML::Node leg_cfg = cfg["leg"];
  const YAML::Node mounts = cfg["mounts"];

  const double coxa_len = require_scalar<double>(leg_cfg, "coxa_length", "leg");
  const double femur_len = require_scalar<double>(leg_cfg, "femur_length", "leg");
  const double tibia_len = require_scalar<double>(leg_cfg, "tibia_length", "leg");

  const YAML::Node front = mounts["l_front"];
  const YAML::Node middle = mounts["l_middle"];

  std::map<std::string, LegSpec> out;
  for (const std::string side : {"l", "r"}) {
    for (const std::string name : {"front", "middle", "rear"}) {
      const YAML::Node& ref = (name == "middle") ? middle : front;
      const std::string ctx = "mounts.l_" + name;
      const double ref_yaw =
          require_scalar<double>(ref, "yaw_deg", ctx) * M_PI / 180.0;
      const double ref_x = require_scalar<double>(ref, "x", ctx);
      const double ref_y = require_scalar<double>(ref, "y", ctx);

      const double x_fr = (name == "rear") ? -ref_x : ref_x;
      const double yaw_fr = (name == "rear") ? (M_PI - ref_yaw) : ref_yaw;
      const double mx = x_fr;
      const double my = (side == "r") ? -ref_y : ref_y;
      const double myaw = (side == "r") ? -yaw_fr : yaw_fr;

      LegSpec spec;
      spec.mount_xyz = Point3(mx, my, 0.0);
      spec.mount_yaw = myaw;
      spec.coxa_len = coxa_len;
      spec.femur_len = femur_len;
      spec.tibia_len = tibia_len;
      out[side + "_" + name] = spec;
    }
  }
  return out;
}

}  // namespace hexa_kinematics
